Optimisation: implement `lcd_print_pad_P()`

gudnimg opened the pull request:
Used `lcd_print_pad_P` is some obvious places (mainly the menu code). We can probably use this function in more places to further save flash memory

Skimming through the menus I didn't see any issues. This needs to be tested well :) 

Change in memory:
Flash: -496 bytes
SRAM: -4 bytes

gudnimg committed:
Optimisations: implement lcd_print_pad_P()

Used lcd_print_pad_P is some obvious places. We can probably use
this function in more places to further save flash memory

Change in memory:
Flash: -268 bytes
SRAM: 0 bytes

leptun commented on Firmware/Marlin_main.cpp:
```diff
@@ -806,7 +806,8 @@ int uart_putchar(char c, FILE *)
 	lcd_clear(); // clears display and homes screen
-	lcd_printf_P(PSTR("\n Original Prusa i3\n   Prusa Research\n%20.20S"), PSTR(FW_VERSION));
+	lcd_puts_P(PSTR("\n Original Prusa i3\n   Prusa Research\n"));
+	lcd_print_pad_P(PSTR(FW_VERSION), LCD_WIDTH);
```
Doesn't this move the text to be left aligned? The old one is right aligned

gudnimg replied:
Good catch. The savings here were only minimal. I'll revert this change.

leptun commented on Firmware/lcd.h:
```diff
@@ -54,6 +54,7 @@ extern void lcd_printFloat(double number, uint8_t digits);
 
 extern void lcd_print(const char*);
 extern char lcd_print_pad(const char* s, uint8_t len);
+char lcd_print_pad_P(const char* s, uint8_t len);
```
why not make this one also extern?

gudnimg replied:
I don't see a reason to use `extern` here. We are including the header file in `.cpp` files this function is used so `extern` is not useful here. We could remove `extern` keyword from most of the functions in the firmware without any issues.

leptun replied:
I agree. I mentioned it just for consistency with the other prototypes around it

leptun commented on Firmware/menu.cpp:
```diff
@@ -197,7 +179,7 @@ static void menu_draw_toggle_puts_P(const char* str, const char* toggle, const u
     sprintf_P(lineStr + LCD_WIDTH - ((settings & 0x02)?strlen_P(toggle):strlen(toggle)) - 3, (settings & 0x02)?PSTR("[%S]%c"):PSTR("[%s]%c"), toggle, eol);
     if (!(settings & 0x04)) lcd_set_cursor(0, menu_row);
-    fputs(lineStr, lcdout);
+    lcd_print_pad(lineStr, LCD_WIDTH);
```
Not sure how useful this is here. lineStr is still padded by the sprintf_P a bit higher. It might be worth rewriting this entire function at some point so it becomes more readable and optimal

gudnimg replied:
I'll re-check this one. I probably only added this because I saw a change in the memory footprint.

gudnimg replied:
Took a shot at some refactoring in [9f70de8](https://github.com/prusa3d/Prusa-Firmware/pull/3873/commits/9f70de8ba09b90f81e5745ec702d309808f20ee3) and saved 106 bytes. We can do a lot more refactoring but it will affect a lot more area in the code which I don't think we should attempt in this PR.

leptun commented on Firmware/menu.cpp:
```diff
@@ -233,7 +215,9 @@ static void menu_draw_item_select_sheet_E(char type_char, const Sheet &sheet)
     menu_format_sheet_select_E(sheet, buffer);
-    lcd_printf_P(PSTR("%c%-18.18s%c"), menu_selection_mark(), buffer.c, type_char);
+    lcd_putc(menu_selection_mark());
+    lcd_print_pad_P(buffer.c, LCD_WIDTH - 2);
```
buffer.c is in ram. Use the non-P version of lcd_print_pad

gudnimg replied:
Fixed [8207561](https://github.com/prusa3d/Prusa-Firmware/pull/3873/commits/82075610040c6fe4ea44ea7834bc7a7027f34e8d)

leptun commented on Firmware/menu.cpp:
```diff
@@ -242,25 +226,19 @@ static void menu_draw_item_puts_E(char type_char, const Sheet &sheet)
     menu_format_sheet_E(sheet, buffer);
-    lcd_printf_P(PSTR("%c%-18.18s%c"), menu_selection_mark(), buffer.c, type_char);
+    lcd_putc(menu_selection_mark());
+    lcd_print_pad_P(buffer.c, LCD_WIDTH - 2);
```
same issue

gudnimg replied:
Fixed [8207561](https://github.com/prusa3d/Prusa-Firmware/pull/3873/commits/82075610040c6fe4ea44ea7834bc7a7027f34e8d)

leptun commented on Firmware/menu.cpp:
```diff
@@ -242,25 +226,19 @@ static void menu_draw_item_puts_E(char type_char, const Sheet &sheet)
     lcd_set_cursor(0, menu_row);
-    lcd_printf_P(PSTR("%c%-.16S "), menu_selection_mark(), str);
+    lcd_putc(menu_selection_mark());
+    lcd_print_pad_P(str, LCD_WIDTH - 3);
```
This usage is not correct. This function needs/uses only the length limiting modifier, not also the padding modifier. With lcd_print_pad_P, the number after the string is pushed to the end of the line, which is not the desired effect. See the MMU filament menus where this is used (afaik, might be wrong on this statement)

gudnimg replied:
You are correct 👍 Interestingly it's not an issue with "Extruder X" items.
![IMG_2832](https://user-images.githubusercontent.com/8218499/212473386-9acf2307-b232-493f-82d3-3df990201081.jpg)

![IMG_2831](https://user-images.githubusercontent.com/8218499/212473397-8ce5e122-3caf-4214-9aa3-4929e43d37e4.jpg)

gudnimg replied:
Issue is now fixed. I changed `lcd_print_pad_P` such that it now returns the number of padded pads (the remains of `len`) It can help us determine the next cursor position.

leptun commented on Firmware/menu.cpp:
```diff
@@ -448,13 +426,9 @@ static void menu_draw_P(char chr, const char* str, int16_t val);
-	lcd_printf_P(menu_fmt_int3, chr, str, spaces, val);
+    char spaces[LCD_WIDTH + 1] = {0};
+    sprintf_P(spaces, menu_fmt_int3, chr, str, spaces, val);
+    lcd_print_pad_P(spaces, LCD_WIDTH);
```
this one is not ok. Go for example to "Settings->Temperature" and you will see how it fails / crashes the firmware.

gudnimg replied:
Woops I see `spaces` appears twice there. Need to think about this one 🤔

gudnimg replied:
I reverted these changes for now.

gudnimg committed:
Refactor menu_draw_toggle_puts_P

The refactoring drops the local buffer lineStr

Change in memory:
Flash: -106 bytes
SRAM: 0 bytes

gudnimg committed:
Fix an issue with menu_draw_item_puts_P

Change in memory:
Flash: +16 bytes
SRAM: 0 bytes

gudnimg committed:
optimisation:  refactor menu_draw_P()

Add a way to only change the LCD column

Change in memory:
Flash: -90 bytes
SRAM: 0 bytes

gudnimg commented:
Difference in memory on Multilang build:

Before:
```
[build] Program:  252682 bytes (96.4% Full)
[build] (.text + .data + .bootloader)
[build] 
[build] Data:       5648 bytes (68.9% Full)
[build] (.data + .bss + .noinit)
```

After:
```
[build] Program:  252228 bytes (96.2% Full)
[build] (.text + .data + .bootloader)
[build] 
[build] Data:       5648 bytes (68.9% Full)
[build] (.data + .bss + .noinit)
```

DRracer commented on Firmware/lcd.cpp:
```diff
@@ -329,13 +329,31 @@ void lcd_no_autoscroll(void)
+/// @return row offset which the LCD register understands
+static uint8_t __attribute__((noinline)) lcd_get_row_offset(uint8_t row)
 {
 	uint8_t row_offsets[] = { 0x00, 0x40, 0x14, 0x54 };
```
Can we move this array into PROGMEM?

gudnimg replied:
Done

gudnimg committed:
optimisation: move row_offsets into PROGMEM

Change in memory:
Flash: -42 bytes
SRAM: -4 bytes

leptun commented on Firmware/lcd.cpp:
```diff
@@ -341,7 +343,6 @@ static void FORCE_INLINE lcd_set_current_row(uint8_t row)
 static uint8_t __attribute__((noinline)) lcd_get_row_offset(uint8_t row)
 {
-	uint8_t row_offsets[] = { 0x00, 0x40, 0x14, 0x54 };
 	return row_offsets[min(row, LCD_HEIGHT - 1)];
```
I'm pretty sure you must use pgm_read_byte for the progmem array

gudnimg replied:
You're right, I forgot about that part.

gudnimg replied:
@leptun Fixed. Flash usage increased by 4 bytes. I'll run this again quickly on my machine later this evening.

gudnimg committed:
Fix an issue with indexing PROGMEM pointer

Using array index 'pointer[index]' doesn't work properly.
Instead using "pointer + index" works fine.

No change in memory footprint

DRracer commented on Firmware/lcd.cpp:
```diff
@@ -329,13 +331,30 @@ void lcd_no_autoscroll(void)
+
+void lcd_set_cursor_column(uint8_t col)
+{
+	lcd_command(LCD_SETDDRAMADDR | (col + lcd_get_row_offset(lcd_currline)));
```
`lcd_command(LCD_SETDDRAMADDR | (col + lcd_get_row_offset(lcd_currline)));`
seems to be duplicated in these functions - I'm not sure if doing something like this actually saves anything:
```
void lcd_set_cursor(uint8_t col, uint8_t row)
{
	lcd_set_current_row(row);
	lcd_set_cursor_column(col);
}

void lcd_set_cursor_column(uint8_t col)
{
	lcd_command(LCD_SETDDRAMADDR | (col + lcd_get_row_offset(lcd_currline)));
}
```

DRracer replied:
yes, it saves another 22B, but `lcd_set_cursor` must be made `__attribute__((noinline))`

gudnimg replied:
The downside here is sending two commands to the LCD. I dont remember how long it took but I think one command transfer is in the order of 20 to 50us.

gudnimg replied:
IMO I would keep this as is for speed. 22B doesnt seem worth it to make the cursor change 2x slower. 

Maybe we can shorten the command to only modify the column. I need to find the datasheet again 🤔

leptun replied:
I would definitely keep the LCD faster at the cost of 22B. It's already painfully slow. Let's not make it even slower.
The `LCD_SETDDRAMADDR` sets both the row and the column at the same time. Actually, the LCD doesn't really have the concept of rows and columns and that's why we need the row offsets. From the perspective of the LCD, the DDRAM is just a block of memory that we write to and `LCD_SETDDRAMADDR` just updates the r/w pointer to that memory block.

DRracer replied:
Ok, what you wrote makes sense, it would be nice to add some kind of explanation why is the code written as it is. Otherwise, I like it :wink:
